### ISSUE TITLE
[Merge] Optimistic merge policy that self corrects its selections if a much better score is available in future or current merge scores are much worse from previous scores

### ISSUE THREAD

**Issue by itiyama** [opensearch-project/OpenSearch#9383]
**Is your feature request related to a problem? Please describe.**
Lucene's [merge selection algorithm](https://github.com/apache/lucene/blob/main/lucene/core/src/java/org/apache/lucene/index/TieredMergePolicy.java#L435) uses a pessimistic merge selection strategy where it looks at the existing state of the segments and selects the best possible merges at the point in time. In reality, the indexer is emitting out segments every X seconds depending on refresh strategy and not merging can be a better strategy than merging segments right away if we could somehow know that new segments would be created soon. Once a segment is selected for merging, the decision is not revisited for that segment ever till it is merged away.

Example
Assume that there is a segment of size 2GB and many small segments, say 15, adding up to 1 GB. The merge selection algorithm, assuming Opensearch default merge settings, will create 2 merges - one for small segments adding up to 1GB and another one including 2GB segment. While [pathological segment merging fix](https://github.com/apache/lucene/pull/900) in lucene attempts to fix this issue, it is based on a static policy of 50% size ratio. It also creates yet another problem, where it never attempts to merge 2GB segment with smaller segments, even when there is no anticipation of more writes e.g. once a time series shard is rolled over.

**Describe the solution you'd like**
A few improvements that can be tried out:
1. Use an optimistic strategy for merge selection algorithm where the merges that have not yet started can be revisited. The [MergeScore](https://github.com/apache/lucene/blob/main/lucene/core/src/java/org/apache/lucene/index/TieredMergePolicy.java#L278) can be stored in the [merge spec](https://github.com/apache/lucene/blob/main/lucene/core/src/java/org/apache/lucene/index/MergePolicy.java#L463) to decide whether to abort a pending merge if new segments or a better score policy is available. This can be a bit hard to do, especially because we may just end up aggressively aborting.
2. The order in which the merges are executed should be based on the best score.
3. Abort policy for merging can also be based on previous scores for that shard instead of a static policy. A simple Exponential moving average can be tried out.

**Comment by itiyama** [comment]
Tagging @shwetathareja @rishabhmaurya for feedback.

**Comment by gashutos** [comment]
Aborting segment merge could leave the stale temporary files created during merge ?

**Comment by itiyama** [comment]
You are right @gashutos . But the recommendation here is not to abort an ongoing merge. It is just abort pending merges when a better strategy is available or not schedule a merge if the score is too far off from previous ones until a time or size condition is met.

**Comment by shwetathareja** [comment]
@itiyama : There was change done in Lucene to merge the tiny segments at the time of refresh/ commit itself - merge_on_refresh & merge_on_commit - https://issues.apache.org/jira/browse/LUCENE-8962

This was enabled by default in OpenSearch here - https://github.com/opensearch-project/OpenSearch/pull/3561

If we delay merging longer, then it can impact search performance with too many small segments. But, I agree if it can simply evaluate pending merges in case better score policy is available and not abort aggressively. This would reduce the no. of merging operations.

**Comment by RS146BIJAY** [comment]
On a similar note on improving overall merge process and performance, I am exploring few other merge policies which can improve search and update performance. I am taking reference from [this paper](https://scholar.harvard.edu/files/stratos/files/dostoevskykv.pdf) which discusses different compaction policies which optimally trade between the I/O cost of updates on one hand and the I/O cost of lookups and storage space on the other.

Will create a RFC shortly discussing details on these merge policies along with performance gains we can get from them.

**Comment by itiyama** [comment]
> @itiyama : There was change done in Lucene to merge the tiny segments at the time of refresh/ commit itself - merge_on_refresh & merge_on_commit - https://issues.apache.org/jira/browse/LUCENE-8962

Yes, this is one of the triggers for merge to just merge the segments created as a result of refresh. This does not really select large segments for merging and yes, it does provide benefits on large core machines, but it does not eliminate the problem completely it only works on segments created as a result of refresh. For example, a small segment will be created every X seconds for a single core machine.

> Will create a RFC shortly discussing details on these merge policies along with performance gains we can get from them.
Sure, will wait for your RFC.
